### DISCUSSION TITLE
Sequential use of smallCaps.ref for the same object omits its boardID in capData Slots

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#10218] [Q&A] [answered]
# Context

I’d like to bring up an issue I encountered while working on test cases running in the `z:acceptance` test phase, specifically when using the [`smallCapsContext`](https://github.com/Agoric/agoric-3-proposals/blob/main/packages/synthetic-chain/src/lib/cliHelper.ts#L166-L186) method, provided by `@agoric/synthetic-chain`, to handle brands object to construct the body of required offers. 

Although this might not be relevant for everyone, it could be useful for anyone facing a similar issue in the future. Here's a summary of the situation.

The `smallCapsContext` implementation is as follows:

```js
export const smallCapsContext = () => {
  const slots = [] as string[]; // XXX global mutable state
  const smallCaps = {
    Nat: (n: number | bigint) => `+${n}`,
    // XXX mutates obj
    ref: (obj: any) => {
      if (obj.ix) return obj.ix;
      const ix = slots.length;
      slots.push(obj.boardID);
      obj.ix = `$${ix}.Alleged: ${obj.iface}`;
      return obj.ix;
    },
  };

  const toCapData = (body: unknown) => {
    const capData = { body: `#${JSON.stringify(body)}`, slots };
    return JSON.stringify(capData);
  };

  return { smallCaps, toCapData };
};
```

# The Problem

When making multiple sequential offers using the same `brand` object, I encountered an error where the `boardID` was not being included in the `capData` `slots` array. 

The issue arises because the `ix` property is added to the `brand` object after the first call to `ref`. As a result, subsequent calls skip the logic that pushes the `boardID` into the `slots` array, since the condition `if (obj.ix)` is satisfied.

It is important to highlight that I have failed to noticed the comment present on the smallCapsContext function that clearly specified that smallCaps.ref  mutates the object.

# Solution

This occurs because the same reference to the `brand` object is used across multiple offers, and the `ix` attribute remains in the object. To resolve this, you can either:

- **Clone the brand object** before passing it to `smallCaps.ref`, ensuring the `ix` attribute does not persist across calls.
- **Delete the `ix` attribute** from the `brand` object before each reference to `smallCaps.ref`.

To illustrate this behavior more clearly, I’ve created a simple test that demonstrates the issue in detail:  
[smallCapsContext.test.js](https://gist.github.com/Jorge-Lopes/9af4318cc3ac6e8bd212033506fe9003#file-smallcapscontext-test-js)

**Comment by dckc** [comment, marked as answer]
`smallCapsContext` is a kludge - it's a partial implementation of `@agoric/marshal`. If you run into limitations, I suggest using the real [toCapData](https://endojs.github.io/endo/interfaces/_endo_marshal.Marshal.html#toCapData) function.

I'm a little surprised that `smallCapsContext` is exported from any package. It was originally written to be used in a constrained context where such limitations aren't relevant. If you're working on acceptance testing, I suggest adding `@deprecated` to `smallCapsContext`.

**Reply by Jorge-Lopes** [reply]
Thank you very much for the feedback.

I have followed your advice and removed the use of `smallCapsContext` from the the acceptance test.
I was using that method to build offers to interact with the KREAd contract, the current implementation is the following:
[kread.js](https://github.com/Agoric/agoric-sdk/pull/10242/files#diff-33acbcbb821810d4d9367d105f67cc46490c937976c3c742db8537bb50d79d55)
